velox/exec/tests/CustomJoinTest.cpp

```diff
@@ -0,0 +1,336 @@
+    if (remainingLimit_ <= inputSize) {
+      finished_ = true;
+    }
+
+    if (remainingLimit_ >= inputSize) {
+      remainingLimit_ -= inputSize;
+      auto output = input_;
+      input_.reset();
+      return output;
+    }
+
+    auto output = std::make_shared<RowVector>(
+        input_->pool(),
+        input_->type(),
+        input_->nulls(),
+        remainingLimit_,
+        input_->children());
+    input_.reset();
+    remainingLimit_ = 0;
+    return output;
+  }
+
+  BlockingReason isBlocked(ContinueFuture* future) {
+    if (numRows_.has_value()) {
+      return BlockingReason::kNotBlocked;
+    }
+
+    auto joinBridge = operatorCtx_->task()->getCustomJoinBridge(
+        operatorCtx_->driverCtx()->splitGroupId, planNodeId());
+    auto numRows = std::dynamic_pointer_cast<CustomJoinBridge>(joinBridge)
+                       ->numRowsOrFuture(future);
+
+    if (!numRows.has_value()) {
+      return BlockingReason::kWaitForJoinBuild;
+    }
+    numRows_ = std::move(numRows);
+    remainingLimit_ = numRows_.value();
+
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() {
+    return finished_ || (noMoreInput_ && input_ == nullptr);
+  }
+
+ private:
+  int32_t remainingLimit_;
+  std::optional<int32_t> numRows_;
+
+  bool finished_{false};
+};
+
+class CustomJoinBridgeTranslator : public Operator::PlanNodeTranslator {
+  std::unique_ptr<Operator>
+  toOperator(DriverCtx* ctx, int32_t id, const core::PlanNodePtr& node) {
+    if (auto joinNode = std::dynamic_pointer_cast<const CustomJoinNode>(node)) {
+      return std::make_unique<CustomJoinProbe>(id, ctx, joinNode);
+    }
+    return nullptr;
+  }
+
+  std::unique_ptr<JoinBridge> toJoinBridge(const core::PlanNodePtr& node) {
+    if (auto joinNode = std::dynamic_pointer_cast<const CustomJoinNode>(node)) {
+      auto joinBridge = std::make_unique<CustomJoinBridge>();
+      return joinBridge;
+    }
+    return nullptr;
+  }
+
+  OperatorSupplier toOperatorSupplier(const core::PlanNodePtr& node) {
+    if (auto joinNode = std::dynamic_pointer_cast<const CustomJoinNode>(node)) {
+      return [joinNode](int32_t operatorId, DriverCtx* ctx) {
+        return std::make_unique<CustomJoinBuild>(operatorId, ctx, joinNode);
+      };
+    }
+    return nullptr;
+  }
+};
+
+/// This test will show the CustomJoinBuild passing count of input rows to
+/// CustomJoinProbe via CustomJoinBridge, then probe will emit corresponding
+/// number of rows, like a Limit operator.
+class CustomJoinTest : public OperatorTestBase {
+ protected:
+  void SetUp() override {
+    OperatorTestBase::SetUp();
+    Operator::registerOperator(std::make_unique<CustomJoinBridgeTranslator>());
+  }
+
+  RowVectorPtr makeSimpleRowVector(vector_size_t size) {
+    return makeRowVector(
+        {makeFlatVector<int32_t>(size, [](auto row) { return row; })});
+  }
+
+  void testCustomJoin(
+      int32_t numThreads,
+      const std::vector<RowVectorPtr>& leftBatch,
```

Comment:
nit: leftBatch -> leftBatches and rightBatches since there can be more than one batch

Reply:
Fixed